palette/src/luma/relative_contrast.rs

```diff
@@ -0,0 +1,71 @@
+
+/// A trait for calculating relative contrast between two colors.
+///
+/// W3C's Web Content Accessibility Guidelines (WCAG) 2.1 suggest a method
+/// to calculate accessible contrast ratios of text and background colors
+/// for those with low vision or color deficiencies, and for contrast of user
+/// interface graphics. The possible range of ratios is from 1:1 to 21:1.
+/// There is a Success Criterion for Contrast (Minimum) and a Success Criterion
+/// for Contrast (Enhanced), SC 1.4.3 and SC 1.4.6 respectively, which are
+/// concerned with text and images of text. SC 1.4.11 is a Success Criterion for
+/// "non-text contrast" such as user interface components and other graphics.
+/// The relative contrast is calculated by `(L1 + 0.05) / (L2 + 0.05)`, where
+/// `L1` is the luminance of the brighter color and `L2` is the luminance of the
+/// darker color both in sRGB linear space. Higher contrast ratio is generally
+/// desireable.
+///
+/// For more details of the guidelines:
+///
+/// [Success Criterion 1.4.3 Contrast (Minimum) (Level AA)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum)
+///
+/// [Success Criterion 1.4.6 Contrast (Enhanced) (Level AAA)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-enhanced)
+///
+/// [Success Criterion 1.4.11 Non-text Contrast (Level AA)](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html)
+pub trait RelativeContrast {
+    /// The return type for the contrast ratio.
+    type Scalar: PartialOrd + FromF64;
+
+    /// Calculate contrast ratio between two colors.
+    fn get_contrast_ratio(&self, other: &Self) -> Self::Scalar;
+    /// Verify the contrast between two colors satisfies SC 1.4.3. Contrast
+    /// is at least 4.5:1 (Level AA).
+    fn is_min_contrast_text(&self, other: &Self) -> bool {
+        self.get_contrast_ratio(other) >= from_f64(4.5)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.3 for large
+    /// text. Contrast is at least 3:1 (Level AA).
+    fn is_min_contrast_large_text(&self, other: &Self) -> bool {
+        self.get_contrast_ratio(other) >= from_f64(3.0)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.6. Contrast
+    /// is at least 7:1 (Level AAA).
+    fn is_enhanced_contrast_text(&self, other: &Self) -> bool {
+        self.get_contrast_ratio(other) >= from_f64(7.0)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.6 for large
+    /// text. Contrast is at least 4.5:1 (Level AAA).
+    fn is_enhanced_contrast_large_text(&self, other: &Self) -> bool {
+        self.is_min_contrast_text(other)
+    }
+    /// Verify the contrast between two colors satisfies SC 1.4.11 for graphical
+    /// objects. Contrast is at least 3:1 (Level AA).
+    fn is_min_contrast_graphics(&self, other: &Self) -> bool {
+        self.is_min_contrast_large_text(other)
+    }
+}
+
+impl<T: IntoColor + Copy> RelativeContrast for T {
+    type Scalar = f32;
```

Comment:
Sorry, I was a bit sloppy when I suggested the blanket implementation. You can make this generic, but I think `IntoColor` defaults to `f32`.

Reply:
What would this look like? I saw that `IntoColor` defaults to `f32` but I'm not sure how to use that to satisfy the trait bounds.

Reply:
The defaults are just what the type parameters take when they are omitted, so all you have to do is to add them to the `impl`:

```rust
impl<C, Wp, T> RelativeContrast for C
where
    C: IntoColor<Wp, T>,
    Wp: WhitePoint,
    T: FloatComponent
```

Then you will have To available for the scalar type.